async_fuse/src/fuse/decoder.rs

```diff
@@ -0,0 +1,339 @@
+//! The implementation of byte slice
+
+use super::abi_marker::FuseAbiData;
+
+use std::ffi::OsStr;
+use std::mem;
+use std::slice;
+
+use log::trace;
+use memchr::memchr;
+
+/// A slice of bytes which is used for parsing FUSE abi data.
+#[derive(Debug)]
+pub struct Decoder<'b> {
+    /// inner bytes
+    bytes: &'b [u8],
+}
+
+/// Types which can be decoded from bytes
+#[allow(single_use_lifetimes)]
+pub trait Decode<'b>: Sized {
+    /// Decode from bytes
+    fn decode(de: &'_ mut Decoder<'b>) -> Result<Self, DecodeError>;
+}
+
+/// The error returned by `Decoder`
+#[derive(Debug, PartialEq, Eq, thiserror::Error)]
+pub enum DecodeError {
+    /// Expected more data
+    #[error("NotEnough")]
+    NotEnough,
+
+    /// The pointer's alignment mismatched with the type
+    #[error("AlignMismatch")]
+    AlignMismatch,
+
+    /// Data is more then expected
+    #[error("TooMuchData")]
+    TooMuchData,
+
+    /// Number overflow during decoding
+    #[error("NumOverflow")]
+    NumOverflow,
+    // #[error("InvalidValue")]
+    // InvalidValue,
+}
+
+/// checks pointer alignment, returns `AlignMismatch` if failed
+#[inline]
+fn check_align<T>(ptr: *const u8) -> Result<(), DecodeError> {
+    let addr = utilities::ptr_to_usize(ptr);
+    let align = mem::align_of::<T>();
+    if addr.wrapping_rem(align) != 0 {
+        trace!(
+            "failed to convert bytes to type {}, \
+                    pointer={:p} is not a multiple of alignment={}",
+            std::any::type_name::<T>(),
+            ptr,
+            align,
+        );
+        return Err(DecodeError::AlignMismatch);
+    }
+    Ok(())
+}
+
+/// checks whether there are enough bytes
+#[inline]
+fn check_size(len: usize, need: usize) -> Result<(), DecodeError> {
+    if len < need {
+        trace!(
+            "no enough bytes to fetch, remaining {} bytes but to fetch {} bytes",
+            len,
+            need,
+        );
+        return Err(DecodeError::NotEnough);
+    }
+    Ok(())
+}
+
+impl<'b> Decoder<'b> {
```

Comment:
The functions in the struct Decoder is all about fetch data from u8 slice, while the decode function is actually defined in the T structure. So I suggest to rename this to something related to buf or slice, for example slice generator?

The current method call example is:
```
SomeTypeDeriveDecode a = decoder.decode();
```

I prefer the following one:
```
SomeTypeDeriveDecode a = SomeTypeDeriveDecode::decode(slice_generator);
```
It's more clear that the decode function is defined in the `SomeTypeDeriveDecode`, and the generator only handles the slice stuff.

Reply:
`Decode`/`Decoder` are just like `Deserialize`/`Deserializer` in `serde`.
<https://docs.rs/serde/1.0.118/serde/de/index.html>

byte_slice is not a slice though there is a slice in it.
slice_generator is not a [`Generator`](https://doc.rust-lang.org/stable/std/ops/trait.Generator.html).
I think both of them are misleading.

Reply:
How about making `fetch_*` free functions which take `&mut &[u8]` as input?